### PR TITLE
Add initial docs for the placeholder service

### PR DESCRIPTION
**Pull request by dualspiral**
[SpongeAPI](https://github.com/SpongePowered/SpongeAPI/pull/2116) | [SpongeCommon](https://github.com/SpongePowered/SpongeCommon/pull/2588) | **SpongeDocs (that's here!)**

Documentation for the Placeholder API. ~~As this hasn't been merged yet, and when it is,~~ This isn't in 7.2 but will be in 7.3. As such, this is a draft. Comments, suggestions, thoughts, etc. are welcome.

I'll undraft it once I release API 7.3.

### CODE REVIEW

**Comment by Spongy** [comment]
A preview for this pull request is available at https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/8b853c4/.

Here are some links to the pages that were modified:

- added: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/8b853c4/plugin/placeholders/creatingtokens.html
- added: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/8b853c4/plugin/placeholders/index.html
- added: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/8b853c4/plugin/placeholders/retrievingtext.html
- modified: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/8b853c4/plugin/index.html

_Since the preview frequently changes, please link to [this comment](https://github.com/SpongePowered/SpongeDocs/pull/834#issuecomment-622950977), not to the direct url to the preview._

File: source/plugin/placeholders/creatingtokens.rst [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+If your placeholder is unable to provide text because there is insufficient information, the placeholder should return
+and empty :javadoc:`Text` and not throw an exception.
+
+Registering Your PlaceholderParser
+==================================
+
+For your parser to be easily accessible to other plugins, it must be registered in the registry. This should be done
+by listening to the :javadoc:`GameRegistryEvent.Register<PlaceholderParser>` event and registering your parsers using 
+the :javadoc:`GameRegistryEvent.Register<PlaceholderParser>#register(T) {register}` method.
```
Comment (ST-DDT) [inline source/plugin/placeholders/creatingtokens.rst:147]: Add a sample listener implementation here.

File: source/plugin/placeholders/retrievingtext.rst [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+=============================
+
+There are two ways to obtain a registered :javadoc:`PlaceholderParser`:
+
+* Directly obtain the parser from the Sponge Registry by calling 
+  ``Sponge.getRegistry().getType(PlaceholderParser.class, id);``; or
+* Obtain the parser via :javadoc:`PlaceholderService#getParser(String)`.
+
+In general, we recommend that the second option is used, as plugin-provided implementations may allow for shorthand 
+tokens to resolve to a parser, such as accepting ``name`` as a tokne, retrieving the parser for ``sponge:name``,
```
Comment (ST-DDT) [inline source/plugin/placeholders/retrievingtext.rst]: Typo `tokne` -> `token`

File: source/plugin/placeholders/creatingtokens.rst [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+contain includes:
+
+* An associated :javadoc:`MessageReceiver`, such as a :javadoc:`Player`
+* An argument string which has been provided by a templating engine
+
+This is the minimum as specified by Sponge. If another plugin replaces the :javadoc:`PlaceholderService`, there may a
+:javadoc:`PlaceholderText` that provides more information.
+
+If your placeholder is unable to provide text because there is insufficient information, the placeholder should return
+and empty :javadoc:`Text` and not throw an exception.
```
Comment (ST-DDT) [inline source/plugin/placeholders/creatingtokens.rst]: Add an example implementation here.

Maybe something like this:

````java
class MyTeamTagPlaceholderParser {

Text parse(PlaceholderText source) {
    return source.getMessageReceiver()
        .filter(Player.class)
        .map(myplugin::getTeamTagForPlayer)
        .orElse(Text.empty);
}
````

**Review by ST-DDT — CHANGES_REQUESTED** [review]
Nice and detailed description.
The parser impl needs code samples to showcase the possible usecases.

**Comment by dualspiral** [comment]
Updated as suggested - thanks for the pointers. I opted to create three examples in the "creation" section, with increasing levels of complexity to show how the different parts of the `PlaceholderText` can be used. The last one is a bit complex but I have tried to explain what it does both in the source code and in the bullet points above so that it can be translated.

File: source/plugin/placeholders/creatingtokens.rst [hunk 4]
```diff
@@ -39,9 +42,154 @@ This is the minimum as specified by Sponge. If another plugin replaces the :java
...
+    public class PlayerBalancePlaceholder implements PlaceholderParser {
+        
+        @Override
+        public String getId() {
+            return "spongedocs:balance"
+        }
+
+        @Override
+        public String getName() {
+            return "Balance Placeholder"
```
Comment (ST-DDT) [inline source/plugin/placeholders/creatingtokens.rst]: Copy and paste error: Should be `"Location Placeholder"` same for the id

File: source/plugin/placeholders/creatingtokens.rst [hunk 5]
```diff
@@ -39,9 +42,154 @@ This is the minimum as specified by Sponge. If another plugin replaces the :java
...
+        public Text parse(PlaceholderText placeholderText) {
+            // First: check the EconomyService is available
+            Optional<EconomyService> optionalEconService = Sponge.getServiceManager()
+                .provide(EconomyService.class);
+            if (optionalEconService.isPresent()) {
+                EconomyService econService = optionalEconService.get();
+
+                // This placeholder only applies to players, so check that
+                Optional<Player> player = placeholderText.getAssociatedReceiver()
+                    .map(x -> x instanceof Player ? (Player) x : null);
```
Comment (ST-DDT) [inline source/plugin/placeholders/creatingtokens.rst]: This should probably use `filter()` instead,

File: source/plugin/placeholders/creatingtokens.rst [hunk 6]
```diff
@@ -39,9 +42,154 @@ This is the minimum as specified by Sponge. If another plugin replaces the :java
...
+                    // Finally, check the account exists, and if so, return the value
+                    Optional<UniqueAccount> account = econService
+                        .getOrCreateAccount(player.get().getUniqueId());
+                    return account.map(x -> Text.of(x.getBalance(currency))).orElse(Text.EMPTY);
+                }
+            }
+
+            return Text.EMPTY;
+        }
+    }
```
Comment (ST-DDT) [inline source/plugin/placeholders/creatingtokens.rst:139]: Not sure about this example. It requires a lot of comments in order to be easily understandable.
Have you considered splitting it into multiple parts that can have a translation each?

**Example:**

> ... Introduction ...
>
> ````java
> String getId() { ... }
>
> String getName() { ... }
>
> Text parse(PlaceholderText) {
>     Optional<EconomyService> optionalEconService = ...
>     if(!optionalEconService.isPresent()) {
>         return Text.EMPTY;
>     }
>     EconomyService econService = optionalEconService.get();
>
>     Optional<UniqueAccount> optionalAccount = placeholderText.getAssociatedReceiver()
>           .map(recv -> getAccount(econService, recv));
>     if (!optionalAccount.isPresent()) {
>         return Text.EMPTY;
>     }
>     UniqueAccount account = optionalAccount.get();
>
>     Currency currency = placeholderText.getArgumentString()
>           .map(arg -> getCurrency(econService, arg))
>           .orElseGet(econService::getDefaultCurrency);
>     ...
> ````
>
> You can derive the `Account` like this:
>
> - `Optional<Account> getAccount(EconomyService, MessageReceiver) { ... }`
>
> And use the argument string to determine the currency like this:
>
> - `Currency getCurrency(EconomyService, String) { ... }`
>
> But you could also define yourself a separator character (such as `|`) to split the argumentString into more parameters, e.g. adding a number format, turning currency symbols on, or anything you need for your parser.

@Inscrutable What do you think?

**Review by ST-DDT — CHANGES_REQUESTED** [review]

File: source/plugin/placeholders/creatingtokens.rst (hunk 6 above)
Comment (dualspiral) [reply source/plugin/placeholders/creatingtokens.rst:139]: There are bullet points above that explain the basic workflow. I don't think it's appropriate to describe the economy service in more detail on these pages. I think the bullet points at the top are enough to explain the workflow. For example:

```
You can derive the Account like this:

*    Optional<Account> getAccount(EconomyService, MessageReceiver) { ... }

And use the argument string to determine the currency like this:

    Currency getCurrency(EconomyService, String) { ... }
```

That really isn't appropriate to talk about here and signposts should be made to the economy service docs instead because it detracts from the topic of this document. I will add _that_ in.

I don't really care for the structure of the code one way or another - but it _is_ meant to be a complex example to show different parts of the system. I see the value for the way you've done it though - it does make it easier to see each "section".

I have also previously been asked to put comments into the code. I know they aren't translatable but I still see value of them being there in addition to the bullet points.

I could add a note somewhere above that the string is freeform and can be used however the placeholder wishes, including for multiple arguments, but I don't really want to prescribe how you'd do that.

File: source/plugin/placeholders/creatingtokens.rst (hunk 4 above)
Comment (dualspiral) [reply source/plugin/placeholders/creatingtokens.rst]: And the class name.

File: source/plugin/placeholders/creatingtokens.rst (hunk 5 above)
Comment (dualspiral) [reply source/plugin/placeholders/creatingtokens.rst]: It's to do with the generic, without doing an "unsafe cast", that ensures we get an `Optional<Player>` instead of an `Optional<MessageReceiver>`.

File: source/plugin/placeholders/creatingtokens.rst (hunk 6 above)
Comment (Inscrutable) [reply source/plugin/placeholders/creatingtokens.rst:139]: *Musing* I wonder if the more complex example might be appropriate for the Sponge Plugin Cookbook instead - which has been sadly neglected for some time, I believe. When discussion starts to reach its tentacles into other parts of the API, it can be difficult to decide where it fits in the docs, and how to classify it. If it's an important example, perhaps it justifies a new page entirely, rather than dissect it for classification purposes - and link where appropriate. 

As for comments in code blocks, do what you have to do - but try to keep it short and simple, and expand in the text blocks if required.

File: source/plugin/placeholders/creatingtokens.rst [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+Implementing PlaceholderParser
+==============================
+
+.. note::
+  :javadoc:`PlaceholderParser {PlaceholderParsers}` are :doc:`Catalog Types<../data/catalog-types>`. As with all
+  Catalog Types, the ID of the parser should be plugin namespaced, and therefore be of the form  
+  ``[pluginid]:[placeholderid]``, and must be unique.
+
+:javadoc:`PlaceholderParser` objects take a :javadoc:`PlaceholderText` object which contains the context of the
```
Comment (Inscrutable) [inline source/plugin/placeholders/creatingtokens.rst]: You seem to have turned every reference to the PlaceholderParser into a javadoc link. Is it just me, or does that seem excessive / unnecessary? Looking at other docs pages, the *first* reference is generally linked.

File: source/plugin/placeholders/creatingtokens.rst [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+register it in the Sponge Registry.
+
+:javadoc:`PlaceholderParser {PlaceholderParsers}` are independent of a plugin's :javadoc:`PlaceholderService`.
+
+Implementing PlaceholderParser
+==============================
+
+.. note::
+  :javadoc:`PlaceholderParser {PlaceholderParsers}` are :doc:`Catalog Types<../data/catalog-types>`. As with all
+  Catalog Types, the ID of the parser should be plugin namespaced, and therefore be of the form
```
Comment (Inscrutable) [inline source/plugin/placeholders/creatingtokens.rst]: It's a bit wordy. We can lose "and therefore be" and not lose any meaning imho.

File: source/plugin/placeholders/implementingservice.rst [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+:javadoc:`PlaceholderText.Builder`.
+
+Token Matching
+~~~~~~~~~~~~~~
+
+The :javadoc:`PlaceholderService#getParser(String)` method attempts to select a :javadoc:`PlaceholderParser` based on 
+the supplied token. If the token matches the ID of a registered parser exactly then the service **must** return that
+parser. 
+
+Otherwise, the service may use whatever sensible criteria it sees fit to match the input with a token. Any
```
Comment (Inscrutable) [inline source/plugin/placeholders/implementingservice.rst]: Is "sensible" a reserved word here, or do you just not want people to be silly? 😜

File: source/plugin/placeholders/retrievingtext.rst [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    org.spongepowered.api.service.placeholder.PlaceholderService
+    org.spongepowered.api.service.placeholder.PlaceholderParser
+    org.spongepowered.api.service.placeholder.PlaceholderText
+    org.spongepowered.api.service.placeholder.PlaceholderText.Builder
+    org.spongepowered.api.text.Text
+    org.spongepowered.api.text.TextRepresentable
+    org.spongepowered.api.text.channel.MessageReceiver
+
+
+Obtaining A PlaceholderParser
```
Comment (Inscrutable) [inline source/plugin/placeholders/retrievingtext.rst]: Title Case does not capitalise trivial words like "a".

File: source/plugin/placeholders/retrievingtext.rst [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+object. :javadoc:`PlaceholderText {PlaceholderTexts}` can be created by using a :javadoc:`PlaceholderText.Builder` 
+obtained from :javadoc:`PlaceholderService#placeholderBuilder()`.
+
+The builder, as a minimum, allows for the following context to be provided:
+
+* The :javadoc:`PlaceholderParser` to use to generate the :javadoc:`Text` (this must be provided)
+* An associated :javadoc:`MessageReceiver`, allowing for the placeholder to modify its output
+* An argument string that the :javadoc:`PlaceholderParser` can parse, often provided by templating engines
+
+If the service is not the Sponge provided service, the builder may be able to provide more context, though this will
```
Comment (Inscrutable) [inline source/plugin/placeholders/retrievingtext.rst]: Can you put a "then" before "the builder" to clarify this a bit? (perhaps lose the first comma too).

**Review by ST-DDT — APPROVED** [review]

File: source/plugin/placeholders/retrievingtext.rst [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+obtained from :javadoc:`PlaceholderService#placeholderBuilder()`.
+
+The builder, as a minimum, allows for the following context to be provided:
+
+* The :javadoc:`PlaceholderParser` to use to generate the :javadoc:`Text` (this must be provided)
+* An associated :javadoc:`MessageReceiver`, allowing for the placeholder to modify its output
+* An argument string that the :javadoc:`PlaceholderParser` can parse, often provided by templating engines
+
+If the service is not the Sponge provided service, the builder may be able to provide more context, though this will 
+require depending on the service providing plugin in question.
```
Comment (Inscrutable) [inline source/plugin/placeholders/retrievingtext.rst]: "on the plugin that provides the service".

**Review by Inscrutable — COMMENTED** [review]
I know it's all WIP. I figured I'd go over it all anyway. You sure like your javadocs.

File: source/plugin/placeholders/creatingtokens.rst [hunk 13]
```diff
@@ -39,9 +42,154 @@ This is the minimum as specified by Sponge. If another plugin replaces the :java
...
+
+Example: Player Location PlaceholderParser
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This :javadoc:`PlaceholderParser` attempts to get the player's location in the world. If used without a 
+:javadoc:`Player` as the associated receiver, returns an empty :javadoc:`Text`.
+
+.. code-block:: java
+
+    public class PlayerBalancePlaceholder implements PlaceholderParser {
```
Comment (NickImpact) [inline source/plugin/placeholders/creatingtokens.rst]: Just noticed this, but the example class here seems to have met the fatal copy/paste. Looks like this one should be focused on player location.

Comment (dualspiral) [reply source/plugin/placeholders/creatingtokens.rst]: That's from an old commit!

File: source/plugin/placeholders/implementingservice.rst (hunk 9 above)
Comment (dualspiral) [reply source/plugin/placeholders/implementingservice.rst]: This whole page has gone now - so no scope for people being silly here any more!

File: source/plugin/placeholders/retrievingtext.rst (hunk 11 above)
Comment (dualspiral) [reply source/plugin/placeholders/retrievingtext.rst]: Service has gone, so has this line!

File: source/plugin/placeholders/retrievingtext.rst (hunk 12 above)
Comment (dualspiral) [reply source/plugin/placeholders/retrievingtext.rst]: Service has gone. It's making it really easy to deal with these comments!

File: source/plugin/placeholders/creatingtokens.rst (hunk 6 above)
Comment (dualspiral) [reply source/plugin/placeholders/creatingtokens.rst:139]: I went for a much simpler example.

File: source/plugin/placeholders/creatingtokens.rst (hunk 7 above)
Comment (dualspiral) [reply source/plugin/placeholders/creatingtokens.rst]: Sorted it.

File: source/plugin/placeholders/creatingtokens.rst [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+  a function that takes a ``PlaceholderContext`` and returns a ``Text``.
+* Directly implement the interface.
+
+.. note::
+  ``PlaceholderParsers`` are :doc:`Catalog Types<../data/catalog-types>`. If you implement the interface directly,
+  remember that the ID of the parser should be plugin namespaced, of the form  ``[pluginid]:[placeholderid]``. IDs 
+  must also be unique.
+
+:javadoc:`PlaceholderParser` objects take a :javadoc:`PlaceholderContext` object which contains the context of the
+request and returns a :javadoc:`Text` based on that context. Information that the ``PlaveholderContext`` may
```
Comment (ST-DDT) [inline source/plugin/placeholders/creatingtokens.rst]:
```suggestion
request and returns a :javadoc:`Text` based on that context. Information that the ``PlaceholderContext`` may 
```

File: source/plugin/placeholders/creatingtokens.rst [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+request and returns a :javadoc:`Text` based on that context. Information that the ``PlaveholderContext`` may 
+contain includes:
+
+* An associated object, such as a :javadoc:`Player`
+* An argument string which generally will be provided by a templating engine
+
+This is the minimum as specified by Sponge. 
+
+If your placeholder is unable to provide text because there is insufficient information, the placeholder should return
+an empty ``Text`` and not throw an exception.
```
Comment (ST-DDT) [inline source/plugin/placeholders/creatingtokens.rst]: Maybe use `insufficient or malformed information`

File: source/plugin/placeholders/creatingtokens.rst [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+                .map(x -> x.getWorldName())
+                .orElse(Text.EMPTY);
+        })
+        .build();
+
+Example: Player Location PlaceholderParser
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This ``PlaceholderParser`` attempts to get the player's location in the world. If used without a ``Player`` as the 
+associated object, it returns an empty :javadoc:`Text`. This implements the ``PlaveholderParser`` interface directly.
```
Comment (ST-DDT) [inline source/plugin/placeholders/creatingtokens.rst]:
```suggestion
associated object, it returns an empty :javadoc:`Text`. This implements the ``PlaceholderParser`` interface directly.
```

File: source/plugin/placeholders/creatingtokens.rst [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+                .orElse(Text.EMPTY);
+        }
+    }
+
+
+Example: Current Time PlaceholderParser
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This ``PlaceholderParser`` returns the current time in the server's local timezone. If the string "UTC" is provided as
+the argument string, it returns the current time in the UTC time zone. This implements the ``PlaveholderParser``
```
Comment (ST-DDT) [inline source/plugin/placeholders/creatingtokens.rst]:
```suggestion
the argument string, it returns the current time in the UTC time zone. This implements the ``PlaceholderParser`` 
```

File: source/plugin/placeholders/creatingtokens.rst [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+the argument string, it returns the current time in the UTC time zone. This implements the ``PlaveholderParser`` 
+interface directly.
+
+.. code-block:: java
+
+    public class PlayerLocationPlaceholder implements PlaceholderParser {
+
+        @Override
+        public Text parse(PlaceholderContext placeholderContext) {
+            if (placeholderContext.getArgumentString().map(x -> x.equalsIgnoreCase("UTC")).isPresent()) {
```
Comment (ST-DDT) [inline source/plugin/placeholders/creatingtokens.rst]:
```suggestion
            if (placeholderContext.getArgumentString().filter(x -> x.equalsIgnoreCase("UTC")).isPresent()) {
```

File: source/plugin/placeholders/creatingtokens.rst [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+        }
+
+        @Override
+        public String getId() {
+            return "spongedocs:currenttime";
+        }
+
+        @Override
+        public String getName() {
+            return "Placeholder Test";
```
Comment (ST-DDT) [inline source/plugin/placeholders/creatingtokens.rst]: Use a better name for this parser,

File: source/plugin/placeholders/creatingtokens.rst [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+This ``PlaceholderParser`` returns the current time in the server's local timezone. If the string "UTC" is provided as
+the argument string, it returns the current time in the UTC time zone. This implements the ``PlaveholderParser`` 
+interface directly.
+
+.. code-block:: java
+
+    public class PlayerLocationPlaceholder implements PlaceholderParser {
+
+        @Override
+        public Text parse(PlaceholderContext placeholderContext) {
```
Comment (ST-DDT) [inline source/plugin/placeholders/creatingtokens.rst]: Please keep the method order the same as the previous example for consistency reasons.

- getId
- getName
- parse

File: source/plugin/placeholders/creatingtokens.rst [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+Example: Current Time PlaceholderParser
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This ``PlaceholderParser`` returns the current time in the server's local timezone. If the string "UTC" is provided as
+the argument string, it returns the current time in the UTC time zone. This implements the ``PlaveholderParser`` 
+interface directly.
+
+.. code-block:: java
+
+    public class PlayerLocationPlaceholder implements PlaceholderParser {
```
Comment (ST-DDT) [inline source/plugin/placeholders/creatingtokens.rst]:
```suggestion
    public class CurrentTimePlaceholder implements PlaceholderParser {
```

File: source/plugin/placeholders/creatingtokens.rst [hunk 22]
```diff
@@ -0,0 +1,159 @@
...
+* Using :javadoc:`PlaceholderParser#builder()`, supplying your :javadoc:`PluginContainer`, un-namespaced ID and 
+  a function that takes a ``PlaceholderContext`` and returns a ``Text``.
+* Directly implement the interface.
+
+.. note::
+  ``PlaceholderParsers`` are :doc:`Catalog Types<../data/catalog-types>`. If you implement the interface directly,
+  remember that the ID of the parser should be plugin namespaced, of the form  ``[pluginid]:[placeholderid]``. IDs 
+  must also be unique.
+
+:javadoc:`PlaceholderParser` objects take a :javadoc:`PlaceholderContext` object which contains the context of the
```
Comment (ST-DDT) [inline source/plugin/placeholders/creatingtokens.rst:38]: Missing import for `PlaceholderContext`

**Review by ST-DDT — CHANGES_REQUESTED** [review]

**Review by ST-DDT — APPROVED** [review]
I haven't tested the actual code, but otherwise looks good to me.
